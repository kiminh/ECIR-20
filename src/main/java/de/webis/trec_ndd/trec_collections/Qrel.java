package de.webis.trec_ndd.trec_collections;

import java.io.Serializable;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

@Data
@NoArgsConstructor
@Accessors(chain = true)
@SuppressWarnings("serial")
public class Qrel implements Serializable{
    private int score;
    private int topicNumber;
    private String documentID;
    
	public Qrel(String line) {
        String[] args = line.split("[\\s\\t]+");
        
        if (args.length == 5) {
        	//prels, like: 1 clueweb09-en0003-55-31884 0 0 1
            topicNumber = Integer.parseInt(args[0]);
            documentID = args[1];
            //FIXME: what is the score for prels?
            score = Integer.MIN_VALUE;
        }
        else {
            topicNumber = Integer.parseInt(args[0]);
            //skipping 1 because it is irrelevant -> has always 0 as value wehen printed
            documentID = args[2];
            score = Integer.parseInt(args[3]);	
        }
    }
	

	
	public String toString() {
		return new StringBuilder().append(topicNumber).append(" 0 ").append(documentID).append(' ').append(score).toString();
	}
	
	@Override
	public int hashCode() {
		return toString().hashCode();
	}
	
	@Override
	public boolean equals(Object o) {
		if(o instanceof Qrel) {
			return this.documentID.equals(((Qrel) o).documentID) && this.score==((Qrel)o).score && this.topicNumber==((Qrel)o).topicNumber;
		}
		else return false;
	}
}

