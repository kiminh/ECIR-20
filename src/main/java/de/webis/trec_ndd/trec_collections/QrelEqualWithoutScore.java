package de.webis.trec_ndd.trec_collections;

import java.io.Serializable;

import lombok.experimental.Accessors;

@Accessors(chain = true)
@SuppressWarnings("serial")
public class QrelEqualWithoutScore extends Qrel implements Serializable{

	public QrelEqualWithoutScore() {
		
	}
	
	public QrelEqualWithoutScore(String line) {
		super(line);
	}
	
	@Override
	public boolean equals(Object o) {
		if(o instanceof Qrel || o instanceof QrelEqualWithoutScore) {
			return this.getDocumentID().equals(((Qrel) o).getDocumentID()) && this.getTopicNumber()==((Qrel)o).getTopicNumber();
		}
		else return false;
	}
	
	@Override
	public int hashCode() {
		return new StringBuilder().append(getTopicNumber()).append(getDocumentID()).toString().hashCode();
	}
	
	@Override
	public String toString() {
		return super.toString();
	}
	
	public QrelEqualWithoutScore copy() {
		QrelEqualWithoutScore ret = new QrelEqualWithoutScore();
		ret.setDocumentID(this.getDocumentID());
		ret.setTopicNumber(this.getTopicNumber());
		ret.setScore(this.getScore());
		
		return ret;
	}
	
}