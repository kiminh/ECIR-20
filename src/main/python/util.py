def all_files_in_hdfs_dir(hdfs_path):
    import os
    import re
    import subprocess

    cmd = 'hdfs dfs -ls ' + hdfs_path
    cmd_output = subprocess.check_output(cmd, shell=True)
    ret = []
    for line in cmd_output.decode('utf-8').strip().split('\n'):
        fields = re.split('\s+', line)
        if len(fields) == 8:
            ret += [fields[-1]]

    return ret

