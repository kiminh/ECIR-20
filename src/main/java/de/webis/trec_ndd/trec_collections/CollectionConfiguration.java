package de.webis.trec_ndd.trec_collections;

import java.io.InputStream;
import java.io.Serializable;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import de.webis.trec_ndd.trec_collections.SharedTask.TrecSharedTask;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.SneakyThrows;

public interface CollectionConfiguration extends Serializable {

	public String getPathToCollection();
	public String getCollectionType();
	public String getDocumentGenerator();
	public List<SharedTask> getSharedTasks();

	@SneakyThrows
	public default Set<String> judgedDocumentIds() {
		Set<String> ret = new HashSet<>();
		
		for(String qrelResource : getQrelResourcesOfAllSharedTasks()) {
			InputStream resource = AnseriniCollectionReader.class.getResourceAsStream(qrelResource);
			ret.addAll(new QrelReader().readJudgedDocumentIdsFromQrels(resource));
		}
		
		return ret;
	}
	
	public default HashSet<String> documentIdsInRunFiles() {
		HashSet<String> ret = new HashSet<>();
		
		for(SharedTask sharedTask: getSharedTasks()) {
			ret.addAll(sharedTask.documentIdsInRunFiles());
		}
		
		return ret;
	}
	
	public default List<String> getQrelResourcesOfAllSharedTasks() {
		return getSharedTasks().stream()
				.map(SharedTask::getQrelResource)
				.collect(Collectors.toList());
	}
	
	@Getter
	@AllArgsConstructor
	public static enum TrecCollections implements CollectionConfiguration {
		CLUEWEB09(
			// compressed 4TB
			"/mnt/ceph/storage/corpora/corpora-thirdparty/corpus-clueweb09/",
			"ClueWeb09Collection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(
				TrecSharedTask.WEB_2009,
				TrecSharedTask.WEB_2010,
				TrecSharedTask.WEB_2011,
				TrecSharedTask.WEB_2012,
				TrecSharedTask.SESSION_2010,
				TrecSharedTask.SESSION_2011,
				TrecSharedTask.SESSION_2012
			))
		),
		CLUEWEB12(
			"/mnt/ceph/storage/data-in-progress/kibi9872/clueweb12/",
			"ClueWeb12Collection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(
					TrecSharedTask.WEB_2013,
					TrecSharedTask.WEB_2014,
					TrecSharedTask.SESSION_2013,
					TrecSharedTask.SESSION_2014
			))
		),
		GOV1(
			// compressed 4.6 GB
			"/mnt/nfs/webis20/corpora/corpora-thirdparty/corpora-trec/corpus-trec-web/DOTGOV/",
			"TrecwebCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Collections.emptyList())
		),
		GOV2(
			// compressed 81GB
			"/mnt/nfs/webis20/corpora/corpora-thirdparty/corpora-trec/corpus-trec-web/DOTGOV2/gov2-corpus/",
			"TrecwebCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(
					TrecSharedTask.TERABYTE_2004,
					TrecSharedTask.TERABYTE_2005_ADHOC,
					TrecSharedTask.TERABYTE_2006
			))
		),
		GOV2_MQ(			// compressed 81GB
				"/mnt/ceph/storage/corpora/corpora-thirdparty/corpora-trec/corpus-trec-web/DOTGOV2/gov2-corpus/",
				"TrecwebCollection",
				"JsoupGenerator",
				Collections.unmodifiableList(Arrays.asList(
					TrecSharedTask.MILLION_QUERY_2007,
					TrecSharedTask.MILLION_QUERY_2008
				))
		),
		ROBUST04(
			"/mnt/ceph/storage/data-in-progress/kibi9872/robust-04/",
			"TrecCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(TrecSharedTask.ROBUST_04))
		),
		CORE2018(
			"/mnt/nfs/webis20/data-in-progress/trec-2018-webis/WashingtonPost/WashingtonPost.v2/data/",
			"WashingtonPostCollection",
			"WapoGenerator",
			Collections.unmodifiableList(Arrays.asList(TrecSharedTask.CORE_2018))
		),
		CORE2017(
			"/mnt/nfs/webis20/data-in-progress/trec-2017-webis/data/nyt/",
			"NewYorkTimesCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(TrecSharedTask.CORE_2017))
		);
		

		private final String pathToCollection,
							collectionType,
							documentGenerator;

		private final List<SharedTask> sharedTasks;
	}
	
	@Getter
	@AllArgsConstructor
	public static enum OtherCollections implements CollectionConfiguration {
		COMMON_CRAWL_2015_02("CommonCrawlCollection", "JsoupGenerator");
		
		private final String collectionType,
			documentGenerator;

		@Override
		public List<SharedTask> getSharedTasks() {
			return Collections.emptyList();
		}

		@Override
		public String getPathToCollection() {
			return null;
		}
	}
}
