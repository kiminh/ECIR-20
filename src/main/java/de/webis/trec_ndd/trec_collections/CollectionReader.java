package de.webis.trec_ndd.trec_collections;

import java.util.List;

public interface CollectionReader {
	public List<CollectionDocument> extractJudgedRawDocumentsFromCollection();
	
	public List<CollectionDocument> extractRunFileDocumentsFromsCollection();

	public List<CollectionDocument> extractRawDocumentsFromCollection();
}
