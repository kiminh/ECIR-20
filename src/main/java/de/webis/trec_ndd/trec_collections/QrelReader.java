package de.webis.trec_ndd.trec_collections;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileReader;
import java.util.Set;

import java.util.List;
import java.util.ArrayList;
import java.util.HashSet;

public class QrelReader {

	public List<Qrel> readFile(String filepath) throws IOException{
		BufferedReader reader = new BufferedReader(new FileReader(filepath));
		String line;
		List<Qrel> results = new ArrayList<Qrel>();
		while((line=reader.readLine())!=null) {
			results.add(new Qrel(line));
		}
		reader.close();
		return results;
	}
	
	public List<QrelEqualWithoutScore> readFileQrelEqualWithoutScore(String filePath) throws IOException{
		return readFileQrelEqualWithoutScore(new File(filePath));
	}
	
	public List<QrelEqualWithoutScore> readFileQrelEqualWithoutScore(File file) throws IOException{
		BufferedReader reader = new BufferedReader(new FileReader(file));
		String line;
		List<QrelEqualWithoutScore> results = new ArrayList<QrelEqualWithoutScore>();
		while((line=reader.readLine())!=null) {
			while(line.contains("  ")) line=line.replaceAll("  "," ");
			results.add(new QrelEqualWithoutScore(line));
		}
		reader.close();
		return results;
		
	}

	public Set<String> readJudgedDocumentIdsFromQrels(File file) throws IOException {
		return readJudgedDocumentIdsFromQrels(new FileInputStream(file));
	}
	
	public Set<String> readJudgedDocumentIdsFromQrels(InputStream inputStream) throws IOException {
		BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream));
		String line;
		Set<String> results = new HashSet<String>();
		while((line=reader.readLine())!=null) {
			results.add(new Qrel(line).getDocumentID());
		}
		reader.close();
		return results;

		/*return new QueryJudgments(qrelsFile)
				.getQrels().values().stream()
				.map(Map::keySet)
				.flatMap(Set::stream)
				.collect(Collectors.toSet());
		*/
	}
}
