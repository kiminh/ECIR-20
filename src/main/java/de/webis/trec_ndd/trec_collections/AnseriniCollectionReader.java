package de.webis.trec_ndd.trec_collections;

import java.io.Serializable;
import java.lang.reflect.InvocationTargetException;
import java.net.URL;
import java.nio.file.Paths;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;

import org.apache.lucene.document.Document;

import com.google.common.collect.Iterators;

import de.webis.trec_ndd.spark.DocumentHash;
import de.webis.trec_ndd.trec_collections.CollectionConfiguration.OtherCollections;
import io.anserini.collection.ClueWeb09Collection;
import io.anserini.collection.ClueWeb12Collection;
import io.anserini.collection.CommonCrawlCollection;
import io.anserini.collection.DocumentCollection;
import io.anserini.collection.Segment;
import io.anserini.collection.SegmentProvider;
import io.anserini.collection.SourceDocument;
import io.anserini.index.IndexCollection;
import io.anserini.index.IndexCollection.Args;
import io.anserini.index.IndexCollection.Counters;
import io.anserini.index.generator.LuceneDocumentGenerator;
import lombok.Data;
import lombok.SneakyThrows;

@Data
@SuppressWarnings("serial")
public class AnseriniCollectionReader<T extends SourceDocument> implements CollectionReader, Serializable {
	private final CollectionConfiguration config;
	
	@Override
	public List<CollectionDocument> extractJudgedRawDocumentsFromCollection() {
		List<CollectionDocument> ret = new LinkedList<>();

		for(String segmentPath : segmentPaths()) {
			judgedCollectionDocumentsInPath(segmentPath).forEachRemaining(ret::add);
		}

		return ret;
	}
	
	@Override
	public List<CollectionDocument> extractRunFileDocumentsFromsCollection() {
		List<CollectionDocument> ret = new LinkedList<>();

		for(String segmentPath : segmentPaths()) {
			documentsFromRunFilesInPath(segmentPath).forEachRemaining(ret::add);
		}

		return ret;
	}

	@Override
	public List<CollectionDocument> extractRawDocumentsFromCollection() {
		List<CollectionDocument> ret = new LinkedList<>();

		for(String segmentPath : segmentPaths()) {
			collectionDocumentsInPath(segmentPath).forEachRemaining(ret::add);
		}

		return ret;
	}

	@SneakyThrows
	public List<String> segmentPaths() {
		return documentCollection().getFileSegmentPaths()
				.stream()
				.map(p -> p.toAbsolutePath().toString())
				.collect(Collectors.toList());
	}
	
	public Iterator<DocumentHash> documentHashIterator(String segmentPath) {
		Iterator<CollectionDocument> iter = collectionDocumentsInPath(segmentPath);
		
		return Iterators.transform(iter, d -> new DocumentHash(d));
	}
	
	public Iterator<CollectionDocument> judgedCollectionDocumentsInPath(String segmentPath) {
		Set<String> judgedDocumentIds = config.judgedDocumentIds();
		
		return Iterators.filter(iteratorForCollectionDocumentsInPath(segmentPath, doc -> judgedDocumentIds.contains(documentId(doc))), i -> i != null);
	}
	
	public Iterator<CollectionDocument> documentsFromRunFilesInPath(String segmentPath) {
		Set<String> documentIdsInRunFiles = config.documentIdsInRunFiles();
		
		return documentsFromRunFilesInPath(segmentPath, documentIdsInRunFiles);
	}
	
	public Iterator<CollectionDocument> documentsFromRunFilesInPath(String segmentPath, Set<String> documentIdsInRunFiles) {
		return Iterators.filter(iteratorForCollectionDocumentsInPath(segmentPath, doc -> documentIdsInRunFiles.contains(documentId(doc))), i -> i != null);
	}
	
	public Iterator<CollectionDocument> collectionDocumentsInPath(String segmentPath) {
		return Iterators.filter(iteratorForCollectionDocumentsInPath(segmentPath, null), i -> i != null);
	}
	
	@SneakyThrows
	private Iterator<CollectionDocument> iteratorForCollectionDocumentsInPath(String segmentPath, Function<Document, Boolean> keepDocument) {
		SegmentProvider<T> collection = documentCollection();
		
		return new AnseriniCollectionSegmentIterator<>(
				collection, 
				documentGenerator(),
				collection.createFileSegment(Paths.get(segmentPath)),
				segmentPath,
				keepDocument
		);
	}

	@Data
	private static class AnseriniCollectionSegmentIterator<T extends SourceDocument>
			implements Iterator<CollectionDocument> {
		private final SegmentProvider<T> collection;
		private final LuceneDocumentGenerator<T> generator;
		private final Segment<T> segment;
		private final String segmentPath;
		private final Function<Document, Boolean> keepDocument;

		@Override
		public boolean hasNext() {
			return segment.hasNext();
		}

		@Override
		public CollectionDocument next() {
			T document = segment.next();
			if (document == null || !document.indexable()) {
				return null;
			}
			
			Document doc = generator.createDocument(document);
			if (doc == null || (keepDocument != null && Boolean.FALSE.equals(keepDocument.apply(doc)))) {
				return null;
			} else {
				CollectionDocument ret = CollectionDocument.fromLuceneDocument(doc);
				ret.setUrl(extractUrlIfPossible(document));
				return ret;
			}
		}
	}
	
	private static URL extractUrlIfPossible(SourceDocument document) {
		try {
			return extractUrlOrFail(document);
		} catch (Exception e) {
			return null;
		}
	}

	private static URL extractUrlOrFail(SourceDocument document) throws Exception {
		if (document instanceof ClueWeb09Collection.Document) {
			return new URL(((ClueWeb09Collection.Document) document).getURL());
		} else if (document instanceof ClueWeb09Collection.Document) {
			return new URL(((ClueWeb12Collection.Document) document).getURL());
		}
		
		return null;
	}
	
	@SneakyThrows
	@SuppressWarnings({ "rawtypes", "unchecked" })
	private SegmentProvider<T> documentCollection()
			throws InstantiationException, IllegalAccessException, ClassNotFoundException {
		SegmentProvider<T> ret = (SegmentProvider) Class.forName("io.anserini.collection." + args().collectionClass)
				.newInstance();
		((DocumentCollection) ret).setCollectionPath(Paths.get(config.getPathToCollection()));

		if(ret instanceof CommonCrawlCollection) {
			((CommonCrawlCollection) ret).setS3Files(((OtherCollections) config).getS3Files());
		}
		
		return ret;
	}

	@SneakyThrows
	@SuppressWarnings({ "rawtypes", "unchecked" })
	private LuceneDocumentGenerator<T> documentGenerator() throws InstantiationException, IllegalAccessException,
			IllegalArgumentException, InvocationTargetException, NoSuchMethodException, SecurityException, Exception {
		Args args = args();
		Class<?> clazz = Class.forName("io.anserini.index.generator." + args.generatorClass);

		return (LuceneDocumentGenerator) clazz.getDeclaredConstructor(Args.class, Counters.class).newInstance(args,
				new IndexCollection(args).new Counters());
	}

	private Args args() {
		Args ret = new Args();
		ret.generatorClass = config.getDocumentGenerator();
		ret.collectionClass = config.getCollectionType();
		ret.index = "-index";
		ret.input = config.getPathToCollection();

		return ret;
	}

	static String documentId(Document doc) {
		return doc.getField(LuceneDocumentGenerator.FIELD_ID).stringValue();
	}
}
