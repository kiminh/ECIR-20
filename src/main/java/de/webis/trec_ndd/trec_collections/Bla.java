package de.webis.trec_ndd.trec_collections;

import java.io.File;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.codehaus.jackson.map.ObjectMapper;

import de.webis.trec_ndd.trec_collections.CollectionConfiguration.TrecCollections;
import lombok.Data;

public class Bla {
	@Data
	public static class BLA {
		private final String topicNumber;
		private final String topicQuery;
		private final Map<String, String> documentToJudgment; 
	}
	
	public static void main(String [] args) throws Exception {
		Map<String, Object> ret = new LinkedHashMap<>();
		
		for(SharedTask task : TrecCollections.CLUEWEB09.getSharedTasks()) {
			ret.put(task.toString(), taskStuff(task));
		}
		
		new ObjectMapper().writeValue(new File("clueweb.json"), ret);
	}
	
	private static Map<String, BLA> taskStuff(SharedTask task) {
		Map<String, BLA> ret = new HashMap<>();
		
		for(Map.Entry<String, Map<String, String>> i : task.documentJudgments().getData().entrySet()) {
			String query = task.getQueryForTopic(i.getKey());
			ret.put(i.getKey(), new BLA(i.getKey(), query, i.getValue()));
		}
		
		return ret;
	}
}
