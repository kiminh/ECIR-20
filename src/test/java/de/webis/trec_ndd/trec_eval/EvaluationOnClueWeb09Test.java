package de.webis.trec_ndd.trec_eval;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.approvaltests.Approvals;
import org.junit.Ignore;
import org.junit.Test;

import avro.shaded.com.google.common.collect.ImmutableMap;
import de.webis.trec_ndd.trec_collections.SharedTask;
import de.webis.trec_ndd.trec_collections.SharedTask.TrecSharedTask;
import lombok.SneakyThrows;

public class EvaluationOnClueWeb09Test {
	
	@Test
	@Ignore
	public void approveBestwatwpRunFromTable2From2009() {
		// watwp 0.043362
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2009, "watwp"));
	}
	
	@Test
	@Ignore
	public void approveWorstMapRunFromTable2From2009() {
		// 
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2009, ""));
	}
	
	@Test
	@Ignore
	public void approveXXXRunFromTable2From2009() {
		// 
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2009, ""));
	}
	
	@Test
	@Ignore
	public void approveXYXRunFromTable2From2009() {
		// 
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2009, ""));
	}
	
	@Test
	public void approveBestMapRunFromTable2From2010() {
		// cmuWiki10 0.157
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2010, "cmuWiki10"));
	}
	
	@Test
	public void approveWorstMapRunFromTable2From2010() {
		// UMa10IASF 0.080
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2010, "UMa10IASF"));
	}
	
	@Test
	public void approveIvoryL2RbFromTable2From2010() {
		// IvoryL2Rb 0.134
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2010, "IvoryL2Rb"));
	}
	
	@Test
	public void approveTHUIR10QaHtRunFromTable2From2010() {
		// THUIR10QaHt 0.128
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2010, "THUIR10QaHt"));
	}
	
	@Test
	public void approveBestERR_20RunFromTable2From2011() {
		// ICTNET11ADR3 0.157
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2011, "ICTNET11ADR3"));
	}
	
	@Test
	public void approveWorstERR_20RunFromTable2From2011() {
		// uwBAadhoc 0.119
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2011, "uwBAadhoc"));
	}
	
	@Test
	public void approvemsrsv2011a3FromTable2From2011() {
		// msrsv2011a3 0.143
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2011, "msrsv2011a3"));
	}
	
	
	@Test
	public void approveDFalah11RunFromTable2From2011() {
		// DFalah11 0.122
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2011, "DFalah11"));
	}
	
	@Test
	public void approveBestERR_20RunFromTable2From2012() {
		// uogTrA44xi 0.313
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2012, "uogTrA44xi"));
	}
	
	@Test
	public void approveWorstERR_20RunFromTable2From2012() {
		// qutwb 0.166
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2012, "qutwb"));
	}
	
	@Test
	public void approveQUTparaBlineFromTable2From2012() {
		// QUTparaBline 0.290
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2012, "QUTparaBline"));
	}
	
	@Test
	public void approveirra12cRunFromTable2From2012() {
		// irra12c 0.290
		Approvals.verifyAsJson(evaluateAllOfficialMeasuresForRun(TrecSharedTask.WEB_2012, "irra12c"));
	}
	
	public static Object evaluateAllOfficialMeasuresForRun(SharedTask sharedTask, String run) {
		Map<String, String> ret = new HashMap<>();
		Path runFile = runFile(run, sharedTask);
		
		for(EvaluationMeasure evalMeasure: sharedTask.getOfficialEvaluationMeasures()) {
			ret.put(evalMeasure.name(), evalSec(evalMeasure, runFile, sharedTask));
		}
		
		return ImmutableMap.<String, Object>builder()
				.put(run, ret)
				.build();
	}
	
	public static Path runFile(String run, SharedTask task) {
		List<String> candidateRunFiles = task.runFiles().stream()
				.filter(p -> p.contains(run))
				.collect(Collectors.toList());
		
		if(candidateRunFiles.size() != 1) {
			throw new RuntimeException("Cant handle run: '"+ run +"' for shared task '"+ task +"'.");
		}
		
		return Paths.get(candidateRunFiles.get(0));
	}
	
	public static String evalSec(EvaluationMeasure measure, Path run, SharedTask task) {
		try {
			return eval(measure, run, task);
		} catch(Exception e) { }
		
		return "EXCEPTION-DURING-COMPUTATION!";
	}
	
	@SneakyThrows
	public static String eval(EvaluationMeasure measure, Path run, SharedTask task) {
		TrecEvaluation eval = new TrecEvaluator().evaluate(run.toFile().toString(), task.getQrelResourceAsFile(), run);
		return String.format("%1$,.3f", measure.evaluate(eval, task));
	}
}
