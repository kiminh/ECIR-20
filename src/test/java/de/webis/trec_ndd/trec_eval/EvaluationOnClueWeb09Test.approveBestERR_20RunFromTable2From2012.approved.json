{
  "uogTrA44xi": {
    "ERR_20": "0.313",
    "NDCG_CUT_20": "0.238",
    "P_20": "0.453",
    "MAP": "0.212"
  }
}